### PR TITLE
[Tutorial] New tutorials for small graphs

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
This is a set of new tutorials ready for new users.  It consists of the following content:

* A Blitz Introduction to DGL - Node Classification (build a 2-layer GCN for Cora)
* How does DGL represent a graph? (about `DGLGraph`)
* Write your own GNN module (about message passing APIs, builtin functions, etc.)
* Link prediction (predicting existence of edges) on full graph (still using Cora)
* Graph classification (2-layer GCN on a GIN dataset, talks about batched graph and readouts)
* Make your own dataset (about `dgl.data.DGLDataset`)

Notes: the tutorials are in a new directory called `new-tutorial`.  After we finish all the new tutorials, the old tutorials should be removed and the `new-tutorial` directory will be renamed.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

### CODE REVIEW

File: new-tutorial/2_dglgraph.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+#    -  For images, consider applying a vision model such as CNNs.
+# 
+#    You can find plenty of materials on how to encode such attributes
+#    into a tensor in the `PyTorch Deep Learning
+#    Tutorials <https://pytorch.org/tutorials/>`__.
+# 
+
+
+######################################################################
+# Graph Queries
```
Comment (jermainewang) [inline new-tutorial/2_dglgraph.py:111]: I feel the phrase may be related to graph DB queries. Maybe "Query graph structures" is better?

File: new-tutorial/2_dglgraph.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+# -------------
+# 
+# ``DGLGraph`` object provides various methods to query a graph structure.
+# 
+
+print(g.num_nodes())
+print(g.num_edges())
+# Out degrees of the internal node
+print(g.out_degrees(0))
+# In degrees of the internal node - note that the graph is directed so the in degree should be 0.
```
Comment (jermainewang) [inline new-tutorial/2_dglgraph.py:121]: internal -> center

File: new-tutorial/2_dglgraph.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+######################################################################
+# Graph Transformations
+# ---------------------
+# 
+
+
+######################################################################
+# You can also transform a graph to another graph.
+# 
+# A common transformation is to create a subgraph from the original graph:
```
Comment (jermainewang) [inline new-tutorial/2_dglgraph.py:134]: DGL provides many APIs to transform a graph to another such as extracting a subgraph from the original one:

File: new-tutorial/2_dglgraph.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+# 
+
+# The original IDs of each node in sg1
+print(sg1.ndata[dgl.NID])
+# The original IDs of each edge in sg1
+print(sg1.edata[dgl.EID])
+# The original IDs of each node in sg2
+print(sg2.ndata[dgl.NID])
+# The original IDs of each edge in sg2
+print(sg2.edata[dgl.EID])
```
Comment (jermainewang) [inline new-tutorial/2_dglgraph.py:156]: Maybe feature extraction is more a common use case?

File: new-tutorial/2_dglgraph.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+dgl.save_graphs('graph.dgl', g)
+dgl.save_graphs('graphs.dgl', [g, sg1, sg2])
+
+# Load graphs
+(g,), _ = dgl.load_graphs('graph.dgl')
+print(g)
+(g, sg1, sg2), _ = dgl.load_graphs('graphs.dgl')
+print(g)
+print(sg1)
+print(sg2)
```
Comment (jermainewang) [inline new-tutorial/2_dglgraph.py:191]: What is `_`?

File: new-tutorial/3_message_passing.py [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+This tutorial assumes that you already know :doc:`the basics of training a
+GNN for node classification <1_introduction>`. However,
+sometimes you would like to invent a new way of aggregating neighbor
+information or a new way of using the aggregation.
+
+By the end of this tutorial you will be able to
+
+-  Understand DGL’s message passing APIs.
+-  Implement GraphSAGE convolution module by your own.
+
```
Comment (jermainewang) [inline new-tutorial/3_message_passing.py:14]: Sometimes, your model goes beyond simply stacking existing GNN modules. For example, you would like to invent a new way of aggregating neighbor information by considering node importance or edge weights. By the end of this tutorial, you will be able to

-  Understand DGL’s message passing APIs.
-  Implement GraphSAGE convolution module by your own.

This tutorial requires knowledge of the basics of training a GNN for node classification <1_introduction>`.

File: new-tutorial/3_message_passing.py [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+import dgl
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+######################################################################
+# This tutorial assumes that you already know the pipeline of full-graph
+# node classification. If not, please refer to the
+# :doc:`introduction <1_introduction>`.
```
Comment (jermainewang) [inline new-tutorial/3_message_passing.py:26]: Consider remove it as you've mentioned it above.

File: new-tutorial/3_message_passing.py [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+# 
+# The following code for data loading and training loop is directly copied
+# from the introduction tutorial.
+#
```
Comment (jermainewang) [inline new-tutorial/3_message_passing.py:30]: I will first talk about how to use message passing APIs and then add this training loop section as FYI.

File: new-tutorial/3_message_passing.py [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+# 
+# .. math::
+# 
+# 
+#    h_v^{(l)} = U^{(l)}\left(h_v^{(l-1)}, m_v^{(l)}\right)
+# 
+# where DGL calls :math:`M^{(l)}` the *message function*, :math:`\sum` the
+# *reduce function* and :math:`U^{(l)}` the *update function*. Note that
+# :math:`\sum` here can represent any function and is not necessarily a
+# summation.
```
Comment (jermainewang) [inline new-tutorial/3_message_passing.py:108]: Shall we use the same notation as in our paper?

File: new-tutorial/3_message_passing.py [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+# function, which gathers and averages the neighbor features. There are
+# three concepts here:
+#
+# * Message function ``fn.copy_u('h', 'm')`` that
+#   copies the node feature under name ``'h'`` as *messages* sent to
+#   neighbors.
+#
+# * Reduce function ``fn.mean('m', 'h_neigh')`` that averages
+#   all the received messages under name ``'m'`` and saves the result as a
+#   new node feature ``'h_N'``.
```
Comment (jermainewang) [inline new-tutorial/3_message_passing.py:185]: You should replace all `h_neigh` with `h_N`.

File: new-tutorial/3_message_passing.py [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+        w : Tensor
+            The edge weight.
+        """
+        with g.local_scope():
+            g.ndata['h'] = h
+            g.edata['w'] = w
+            g.update_all(message_func=fn.u_mul_e('h', 'w', 'm'), reduce_func=fn.mean('m', 'h_neigh'))
+            h_neigh = g.ndata['h_neigh']
+            h_total = torch.cat([h, h_neigh], dim=1)
+            return self.linear(h_total)
```
Comment (jermainewang) [inline new-tutorial/3_message_passing.py:264]: Same here. Replace `h_neigh` with `h_N`.

File: new-tutorial/3_message_passing.py [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+# the incoming messages:
+# 
+
+def sum_udf(nodes):
+    return {'h': nodes.mailbox['m'].sum(1)}
+
+
+######################################################################
+# In short, DGL will group the nodes by their in-degrees, and for each
+# group DGL stacks the incoming messages along the second dimension. One
```
Comment (jermainewang) [inline new-tutorial/3_message_passing.py:321]: You can ?

**Review by jermainewang — COMMENTED** [review]
Comments for tutorials 1-3

**Comment by BarclayII** [comment]
@jermainewang If you finished reviewing please let me know.

File: new-tutorial/5_graph_classification.py [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+######################################################################
+# Define Model
+# ------------
+# 
+# This tutorial will build a two-layer `Graph Convolutional Network
+# (GCN) <http://tkipf.github.io/graph-convolutional-networks/>`__. Each of
+# its layer computes new node representations by aggregating neighbor
+# information. If you have gone through the
+# :doc:`introduction <1_introduction>`, you will need to notice two
+# differences:
```
Comment (jermainewang) [inline new-tutorial/5_graph_classification.py:150]: you will ~~need to~~ notice two differences

File: new-tutorial/5_graph_classification.py [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+# 
+# -  Since you will be predicting a single category for the *entire graph*
```
Comment (jermainewang) [inline new-tutorial/5_graph_classification.py:152]: Since the task is to predict ...

File: new-tutorial/5_graph_classification.py [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+
+
+
+######################################################################
+# Training Loop
+# -------------
+# 
+# Training loop simply involves iterating over the training set with the
+# ``DataLoader`` you created and computing the gradients, just as you
+# would do for image classification or language modeling.
```
Comment (jermainewang) [inline new-tutorial/5_graph_classification.py:188]: The training loop iterates over the training set with the
``DataLoader`` object and computes the gradients, just like image classification or language modeling.

File: new-tutorial/6_load_data.py [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+
+"""
+
+
+######################################################################
+# ``DGLDataset`` Object Overview
+# ------------------------------
+# 
+# Your custom graph dataset should subclass the ``dgl.data.DGLDataset``
+# object and implement the following methods:
```
Comment (jermainewang) [inline new-tutorial/6_load_data.py:22]: ... the ``dgl.data.DGLDataset`` class ...

File: new-tutorial/6_load_data.py [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+# .. note::
+# 
+#    The original Zachary’s Karate Club network does not have
+#    member ages. The ages in this tutorial are generated synthetically
+#    for demonstrating how to add node features into the graph for dataset
+#    creation.
+# 
+# .. note::
+# 
+#    In practice taking age directly as a numeric feature may
```
Comment (jermainewang) [inline new-tutorial/6_load_data.py:62]: In practice, taking age ...

**Comment by jermainewang** [comment]
@BarclayII I've finished my review.

File: new-tutorial/2_dglgraph.py (hunk 5 above)
Comment (BarclayII) [reply new-tutorial/2_dglgraph.py:191]: `dgl.load_graphs` normally returns a list of graphs and a dict of labels.  If `save_graphs` did not store any label, `dgl.load_graphs` will return an empty dict instead.  Hence the `_`.

File: new-tutorial/2_dglgraph.py [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+#    from one node to the other are often different between both
+#    directions. If you want to handle undirected graphs, you may consider
+#    treating it as a bidirectional graph. See `Graph
+#    Transformations`_ for an example of making
+#    a bidirectional graph.
+# 
+
+
+######################################################################
+# Assigning Node and Edge Features to Graph
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply new-tutorial/2_dglgraph.py:65]: "Assigning" is better since present participles can represent nouns.  I'm scanning other places and changing other verbs that don't follow this.

File: new-tutorial/3_message_passing.py (hunk 9 above)
Comment (BarclayII) [reply new-tutorial/3_message_passing.py:108]: The notations are from the MPNN paper.  Since we are referencing their paper I prefer this notation.

File: new-tutorial/2_dglgraph.py (hunk 4 above)
Comment (BarclayII) [reply new-tutorial/2_dglgraph.py:155]: Added below.

File: new-tutorial/4_link_predict.py [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+
+# Create testing set.
+test_u = torch.cat([torch.as_tensor(test_pos_u), torch.as_tensor(test_neg_u)])
+test_v = torch.cat([torch.as_tensor(test_pos_v), torch.as_tensor(test_neg_v)])
+test_label = torch.cat([torch.zeros(len(test_pos_u)), torch.ones(len(test_neg_u))])
+
+
+######################################################################
+# When training, you will need to remove the edges in the test set from
+# the original graph. You can do this via ``dgl.remove_edges``.
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply new-tutorial/4_link_predict.py:111]: Done.

**Comment by BarclayII** [comment]
@jermainewang Would you mind taking another look on graph classification in particular?  @hetong007 made changes in the tutorial to use `GraphDataLoader`.

File: new-tutorial/5_graph_classification.py [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+test_sampler = SubsetRandomSampler(torch.arange(num_train, num_examples))
+
+train_dataloader = GraphDataLoader(
+    dataset, sampler=train_sampler, batch_size=5, drop_last=False)
+test_dataloader = GraphDataLoader(
+    dataset, sampler=test_sampler, batch_size=5, drop_last=False)
+
+
+######################################################################
+# You can try to iterate over the created ``DataLoader`` and see what it
```
Comment (hetong007) [inline new-tutorial/5_graph_classification.py:88]: ``DataLoader`` -> ``GraphDataLoader``

File: new-tutorial/5_graph_classification.py [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+# 
+# -  Since the task is to predict a single category for the *entire graph*
+#    instead of for every node, you will need to aggregate the
+#    representations of all the nodes and potentially the edges to form a
+#    graph-level representation. Such process is more commonly referred as
+#    a *readout*. A simple choice is to average the node features of a
+#    graph with ``dgl.mean_nodes()``.
+#
+# -  The input graph to the model will be a batched graph yielded by the
+#    ``DataLoader``. The readout functions provided by DGL can handle
```
Comment (hetong007) [inline new-tutorial/5_graph_classification.py:147]: ditto

**Comment by jermainewang** [comment]
@BarclayII  I'm good with the new graph classification tutorial.
